core/src/main/java/org/infinispan/topology/DefaultRebalancePolicy.java

```diff
@@ -84,15 +85,21 @@ public void updateCacheStatus(String cacheName, ClusterCacheStatus cacheStatus)
       clusterTopologyManager.triggerRebalance(cacheName);
    }
 
-   public boolean isBalanced(ConsistentHash ch) {
-      int numSegments = ch.getNumSegments();
-      for (int i = 0; i < numSegments; i++) {
-         int actualNumOwners = Math.min(ch.getMembers().size(), ch.getNumOwners());
-         if (ch.locateOwnersForSegment(i).size() != actualNumOwners) {
-            return false;
-         }
-      }
-      return true;
+   public boolean isBalanced(ClusterCacheStatus cacheStatus) {
+       if(cacheStatus == null){
+           return true;
+       }
+       synchronized(cacheStatus){
+           CacheTopology cacheTopology = cacheStatus.getCacheTopology();
+           ConsistentHashFactory chFactory = cacheStatus.getJoinInfo().getConsistentHashFactory();
+           ConsistentHash currentCH = cacheTopology.getCurrentCH();
+           ConsistentHash updatedMembersCH = chFactory.updateMembers(currentCH, cacheStatus.getMembers());
+           ConsistentHash balancedCH = chFactory.rebalance(updatedMembersCH);
+           if (currentCH == null || balancedCH.equals(currentCH)) {
```

Comment:
If `currentCH == null`, `updateMembers(currentCH, ...)` would throw a NPE. So we should either check for `null` before, or remove the check here. I don't think it's possible to have a `CacheTopology` with a `currentCH == null`, is it?